### ISSUE TITLE
Is it acceptable to implicitly suggest implementing version control for a VDR?

### ISSUE THREAD

**Issue by shigeya** [w3c/did-core#580]
While the draft states the implementation details of the methods is out of scope, some DID  document metadata property, namely, `nextUpdate`, `nextVersionId`, and possibly `versionId` suggest implementing version management in VDR.
(These are SHOULD properties)

Since these properties are added relatively recently, I want to check the above slight inconsistency is acceptable among members of DID WG.

Note: I'm not sure whether there is any security or privacy concern yet.

**Comment by peacekeeper** [comment]
The original idea of this was that it would be completely up to the DID method. There are many DID methods that can't or don't want to support this.

Considering this, I think personally I would prefer MAY over SHOULD for these metadata properties, in my opinion that would be closer to the original intention.

The related DID parameters `versionId` and `versionTime` say that support for them is OPTIONAL.

**Comment by peacekeeper** [comment]
@oed what do you think?

**Comment by shigeya** [comment]
> Considering this, I think personally I would prefer MAY over SHOULD for these metadata properties, in my opinion that would be closer to the original intention.
> 
> The related DID parameters `versionId` and `versionTime` say that support for them is OPTIONAL.

+1. Personal opinion too.

**Comment by oed** [comment]
Using MAY is probably fine. Might be worth noting that methods that doesn't implement this are much less secure.

**Comment by shigeya** [comment]
> Using MAY is probably fine.

Good, thank you.

> Might be worth noting that methods that don't implement these are much less secure.

Before I submit this issue, I've briefly read #483. I understand your concern and the merged proposal.

We need some text for a note if we want to include that warning. Just mentioning `much less secure` does not provide enough information to an implementer. Any suggestions?

**Comment by oed** [comment]
Thanks @shigeya,

How about something like this for `nextUpdate`, and maybe `nextVersionId`.
> This property allows applications to verify if and when the DID Document has changed. This could be critical information if the application needs to verify that a signature made at a specific point in time is valid. Therefore, certain applications might require the use of these properties.

**Comment by shigeya** [comment]
@oed Thank you. Looks good for me. I prefer to add the following text:

> Thus, implementing this property is recommended.

@peacekeeper Do we want to include the above texts as a note?

**Comment by peacekeeper** [comment]
@oed and @shigeya I think it makes sense to add such language, but perhaps you would also consider creating two separate PRs:

- One PR that only changes SHOULD to MAY in the actual sections where those metadata properties are defined.
- Another PR that adds a new section under "Security Considerations" to describe why and how these properties can improve security.

I would do this because as @shigeya already mentioned above, simply saying that these properties "make applications more secure" is not very concrete. I think that the additional text suggested by @oed above is not sufficient either. How exactly do you verify that "a signature made at a specific point in time is valid"? There is a mega-thread at https://github.com/w3c/did-core/issues/386 that illustrates how complex this topic can be. There is also a PR now at https://github.com/w3c/did-core/pull/569. Maybe you want to contribute to that PR, or do one that builds on it?

**Comment by oed** [comment]
@peacekeeper Agree that it's a complex topic and should probably be addressed mostly in another section. Would be good to have something next to the params though about why they are beneficial and recommended.

**Comment by shigeya** [comment]
@peacekeeper Agreed. I'm going to create the first PR.

**Comment by shigeya** [comment]
@oed Forgot to mention about current schedule towards Candidate Recommendation. The deadline for sending PR is February 9th. So, please kindly send the PR soon.

**Comment by peacekeeper** [comment]
> Would be good to have something next to the params though about why they are beneficial and recommended.

@oed I think that's totally fine on a high level. Sorry, I wasn't really against adding your text suggestion, but I was worried that it could trigger a lot of discussion like in one of the other threads I linked, if you talk about how to verify signatures.

For comparison, the sections about metadata properties [equivalentId](https://w3c.github.io/did-core/#equivalentid) and [canonicalId](https://w3c.github.io/did-core/#canonicalid) also contain some text on how to use them. And then they also have an extra section in Security Considerations called [Equivalence Properties](https://w3c.github.io/did-core/#equivalence-properties) that go into more detail.

**Comment by msporny** [comment]
PR #582 addresses this issue, which will be closed once that PR is merged.

**Comment by msporny** [comment]
PR #582 has been merged, closing.
